numpyro/util.py

```diff
@@ -201,58 +202,64 @@ def progress_bar_factory(num_samples, num_chains):
 
     remainder = num_samples % print_rate
 
+    idx_counter = 0
     tqdm_bars = {}
-    finished_chains = []
+    # lock serializes access to idx_counter
+    lock = Lock()
     for chain in range(num_chains):
         tqdm_bars[chain] = tqdm_auto(range(num_samples), position=chain)
         tqdm_bars[chain].set_description("Compiling.. ", refresh=True)
 
-    def _update_tqdm(arg, transform, device):
-        chain_match = _CHAIN_RE.search(str(device))
-        assert chain_match
-        chain = int(chain_match.group())
+    # Uses resource counting to assign chain ids
+    def _calc_chain_idx(iter_num):
+        nonlocal idx_counter
+        with lock:
+            idx = idx_counter
+            idx_counter += 1
+        return idx
+
+    def _update_tqdm(iter_num, increment, chain):
+        iter_num = int(iter_num)
+        increment = int(increment)
+        chain = int(chain)
+        if iter_num == 0:
```

Comment:
because iter_num is not used, it is clearer to use the assertion `chain == -1`

Reply:
agreed - done